### PR TITLE
support for `external-dns` as `addon` for all three cloud providers

### LINKED ISSUE #222: Add an add-on to install and configure external-dns in all clusters
Add an add-on to install and configure external-dns in all clusters.
The add-on should allow users to configure DNS names for their applications and TSB infrastructure (vmgateway) in custom subzones or the existing shared zone.
As an optional add-on, it must not be coupled to the internals of the Management Plane deployment.

### PR DESCRIPTION
**Pull request by smarunich**
1. The sub-domain `$cluster_name.$external_dns_$cloud_dns_zone` for example `*.aks-r161rc0p3-eastus-0.azure.sandbox.tetrate.io`
2. The credentials for service account are done under the `external-dns` module itself
3. Support for all three cloud providers
4. external-dns is implemented as addon

### CODE REVIEW

File: infra/gcp/variables.tf
Comment (nacx) [inline infra/gcp/variables.tf]: I'd leave only `service` until we address the issue to help customize this. This will get in the way of T1 and T2, and for our demo environments it is easy enough for us to just label the `install.Ingress` resources. WDYT?

File: infra/gcp/main.tf
Comment (nacx) [inline infra/gcp/main.tf]: [nit] let's unify code and use coalesce here as well?

File: infra/gcp/main.tf
Comment (nacx) [inline infra/gcp/main.tf]:
```suggestion
  external_dns_enabled       = var.external_dns_enabled
```

File: modules/gcp/external-dns/main.tf
Comment (nacx) [inline modules/gcp/external-dns/main.tf]: How will this work? Are users supposed to create their DNS names with the cluster name prefix?

File: infra/gcp/main.tf
Comment (smarunich) [reply infra/gcp/main.tf]: done, updated few more of iterations.

File: infra/gcp/main.tf
Comment (smarunich) [reply infra/gcp/main.tf]: lol, indeed sunday night sideeffects :), fixed.

**Comment by nacx** [comment]
I'm not sure about the overall structure. External DNS is clearly an add-on. Actually, it is an even more clear add-on than Elasticsearch or the rate limiting server, and the structure should clearly reflect that and that external-dns should be moved to the addons folder.
I mean, this structure seems to favour convenience versus correctness, and that will lead to make the overall repo harder to reason about and potentially evolve, as we start coupling things.

File: infra/gcp/variables.tf
Comment (smarunich) [reply infra/gcp/variables.tf]: yeah, fair enough.

File: modules/gcp/external-dns/main.tf
Comment (smarunich) [reply modules/gcp/external-dns/main.tf]: at this stage, I have tested public zones as `gcp.sandbox.tetrate.io` and `azure.sandbox.tetrate.io`, i am doing `aws.sandbox.tetrate.io`, i have put the placeholders to support private zones, I think we should sort out private zones  management before merging this PR.

**Comment by smarunich** [comment]
> I'm not sure about the overall structure. External DNS is clearly an add-on. Actually, it is an even more clear add-on than Elasticsearch or the rate limiting server, and the structure should clearly reflect that and that external-dns should be moved to the addons folder. I mean, this structure seems to favour convenience versus correctness, and that will lead to make the overall repo harder to reason about and potentially evolve, as we start coupling things.

agree, my bad,  I have moved the `external-dns` modules  under `addons`.

**Comment by smarunich** [comment]
added support for AWS
added interval for external-dns refresh as a variable, defaults to `5s` 
added tags

File: infra/aws/main.tf
Comment (nacx) [inline infra/aws/main.tf]: This still doesn't feel right. I mean, if it is an _optional addon_, it shouldn't be part fo the main infra module. I expect consumers to be able to do a `make external-dns` the same way they do a `make monitoring` or `make argocd`.
By coupling this here we are just dealing with optional add-ons in two different ways. On the one hand, we have add-ons that appear in the main infra module and you control them with a flag, on the other hand, we have proper add-ons that you explicitly install. This just causes confusion to contributors like me as there is no clear way of adding optional components and makes it hard to evolve the entire thing as now the main modules are coupled to ext-dns no matter what.

I really think this shouldn't be here and external-dns be installed with its dedicate make target like argo or the monitoring stack. This looks more that we're trading developer convenience (at the time of writing the PR) for correctness of the repo structure and that's only going to make things harder to maintain and evolve in the future.

Comment (smarunich) [reply infra/aws/main.tf]: I hear you, as I have been also considering below, so in a way looked to DNS as a required infra...

The only reason it is still there, it is desire to get rid of "register_fqdn" step and perhaps have a common approach for dns records management... and eventually have CNAME for UI access... i.e. make it as baseline component, like we have cert-manager.

We need DNS for:

1. TSB MP endpoint (includes XCP)
2. VM onboarding
3. GWs

Even though we can have a separate pointer to external-dns install, I  will have it done.

Comment (nacx) [reply infra/aws/main.tf]: That makes sense

File: modules/addons/aws/external-dns/main.tf
Comment (nacx) [inline modules/addons/aws/external-dns/main.tf:18]: For my understanding, why do you need this? Isn't it enough to have the custom DNS zone?

Comment (nacx) [reply modules/addons/aws/external-dns/main.tf:18]: Discussed offline. There are limitations in the DNS names we can assign with this model, as users will be required to include the cluster names in the DNS names, but I think it's ok to merge this and iterate on it once merged.

**Review by nacx — DISMISSED** [review]

File: infra/aws/main.tf
Comment (nacx) [inline infra/aws/main.tf]: Here and in the others, what about not propagating the `enabled` flag inside, and just use it as a `count` in this root module declaration? You control here at once if the module is applied or not, instead of having to "count" on every single resource inside the extdns module.

Comment (smarunich) [reply infra/aws/main.tf]: as we do have helm provider configuration under the module... https://github.com/tetrateio/tetrate-service-bridge-sandbox/blob/6cc9d2410b3f7f7944665733fa9d5768aa996860/modules/addons/azure/external-dns/main.tf#L47 

in a way it goes to the point of externalizing... provider configuration away from module...

```
╷
│ Error: Module is incompatible with count, for_each, and depends_on
│
│   on main.tf line 55, in module "external_dns":
│   55:   count                      =  var.external_dns_enabled == true ? 1 : 0
│
│ The module at module.external_dns is a legacy module which contains its own local provider configurations, and so calls to it may not use the count, for_each, or depends_on arguments.
│
│ If you also control the module "../../modules/addons/azure/external-dns", consider updating this module to instead expect provider configurations to be passed by its caller.
╵
╷
```

Comment (nacx) [reply infra/aws/main.tf]: ok, not a big deal then, let's keep it as-is

File: Makefile
Comment (nacx) [inline Makefile:221]: Why is the `init` needed on a destroy phase?

File: Makefile
Comment (nacx) [inline Makefile:246]: Can you explain why the external-dns needs to be removed manually in this step?

File: modules/addons/aws/external-dns/main.tf
Comment (nacx) [inline modules/addons/aws/external-dns/main.tf]: These filters should always be there probably, otherwise we get unexpected random domain names registered an will pollute the DNS.
Let's have it fail if the cluster name can't be read instead of silently defaulting to no domain filters.

File: modules/addons/aws/external-dns/main.tf
Comment (nacx) [inline modules/addons/aws/external-dns/main.tf:6]: This will create the zone according to the `dns_zone` configured in the variables, but the two resources below only take into account when that zone is in the "shared zone".

Now teams have their own accounts. For example, in the MP team, we've been assigned the `mp.aws.tetrate.com` zone in the MP AWS account. With the current code, the cluster subzone that will be created there will not work.

I think we should get rid of the distinction between shared/private, and do all this stuff always, assuming that each account will already have a properly configured DNS zone, which is where we're going.

File: modules/aws/jumpbox/aws_cleanup.sh.tmpl
Comment (nacx) [inline modules/aws/jumpbox/aws_cleanup.sh.tmpl]: Shouldn't we defer to terraform and not the script the deletion of the zone it created?

File: modules/azure/k8s/outputs.tf
Comment (nacx) [inline modules/azure/k8s/outputs.tf:44]: I see no script to cleanup records created by external-dns in Azure. Are they cleaned up?

File: modules/aws/jumpbox/aws_cleanup.sh.tmpl
Comment (nacx) [reply modules/aws/jumpbox/aws_cleanup.sh.tmpl]: Also, if users are creating records inside the `cluster_zone` (and we can have external-dns only watch for those cluster-zone records), then just deleting the zone via terraform would be enough, correct?

The rest will be handled by the register_fqdn, as all other records should be scoped to the cluster zone and then this script wouldn't be neeed.

File: modules/gcp/jumpbox/gcp-cleanup.sh
Comment (nacx) [inline modules/gcp/jumpbox/gcp-cleanup.sh]: Do we really need this?
If this about destroying the cluster zones, why simply destroying the zone via terraform is not enough?

File: Makefile
Comment (gonzaloserrano) [reply Makefile:221]: this bit me last week, when doing destroy it complained of a "missing register_fqdn module" or something similar. Adding this init fixed it.

Comment (gonzaloserrano) [reply Makefile:221]:
```
$ make destroy
╷
│ Error: Module not installed
│
│   on main.tf line 8:
│    8: module "register_fqdn" {
│
│ This module is not yet installed. Run "terraform init" to install all modules required by this configuration.
╵
make: *** [destroy_remote] Error 1
```

File: Makefile
Comment (smarunich) [reply Makefile:246]: I will move the external-dns to the separate execution runtime, like we have argocd and other addons, otherwise it is a chicken and egg problem, where the k8s cluster can be destroyed prior helm chart deletion.

File: modules/addons/aws/external-dns/main.tf
Comment (smarunich) [reply modules/addons/aws/external-dns/main.tf]: sounds good.

File: modules/aws/jumpbox/aws_cleanup.sh.tmpl
Comment (smarunich) [reply modules/aws/jumpbox/aws_cleanup.sh.tmpl]: unfortunately, only azure is able to tolerate the zone deletion with the existing records, however GCP and AWS deletion will fail, if there are any DNS records defined... as DNS records managed out of terraform, we have to have a similar hack :| for clean up applied.

When zone will be removed by the script, terraform will be able to discover the deletion.

File: modules/azure/k8s/outputs.tf
Comment (smarunich) [reply modules/azure/k8s/outputs.tf:44]: yes, azure apis do allow zone termination with the existing records.

File: modules/gcp/jumpbox/gcp-cleanup.sh
Comment (smarunich) [reply modules/gcp/jumpbox/gcp-cleanup.sh]: GCP APIs don't allow the zone deletion with the existing DNS records.

**Comment by smarunich** [comment]
Updated to have `external-dns` addon respect the rest of addons framework, like for example `argocd`, but `external-dns` requires proper destroy as modules do orchestrate a cloud-based infrastructure elements, so added a `destroy_external_dns` target as well.

File: addons/aws/external-dns/variables.tf
Comment (nacx) [inline addons/aws/external-dns/variables.tf]: Let's make this a mandatory parameter. Effectively next week the shared accounts are no longer allowed and we should have users explicitly configure that, or take it from the terraform status, but we can't set a default here.

File: addons/aws/external-dns/variables.tf
Comment (nacx) [inline addons/aws/external-dns/variables.tf]: These two should be mandatory and not provided by default. We need to enforce proper ownership tagging. Everyone should configure their team and owner name.

File: addons/azure/external-dns/variables.tf
Comment (nacx) [inline addons/azure/external-dns/variables.tf]: Same here. Let's just remove these defaults.

File: addons/gcp/external-dns/variables.tf
Comment (nacx) [inline addons/gcp/external-dns/variables.tf]: As with other providers, remove the defaults

File: modules/addons/aws/external-dns/external-dns_aws_cleanup.sh.tmpl
Comment (nacx) [inline modules/addons/aws/external-dns/external-dns_aws_cleanup.sh.tmpl:4]: Should we `set -e` to make sure this returns early with a non-zero code if something goes wrong?

File: modules/aws/k8s/main.tf
Comment (nacx) [inline modules/aws/k8s/main.tf]:
```suggestion
```

**Review by nacx — COMMENTED** [review]
Added some final comments. Overall looks good.

Another thing to consider is adding a note in the `addons/README` explaining that when using the `external-dns` addon you must include the clustername as the DNS prefix when annotating k8s services, otherwise it will not work.

File: modules/addons/aws/external-dns/external-dns_aws_cleanup.sh.tmpl
Comment (smarunich) [reply modules/addons/aws/external-dns/external-dns_aws_cleanup.sh.tmpl:4]: agree, resolved.

File: addons/gcp/external-dns/variables.tf
Comment (smarunich) [reply addons/gcp/external-dns/variables.tf]: agree, resolved.

File: addons/aws/external-dns/variables.tf
Comment (smarunich) [reply addons/aws/external-dns/variables.tf]: agree, resolved.

**Review by nacx — APPROVED** [review]
Nice!

**Pull request merged**
